File: Android/Solution/Exercise3_part2/app/src/main/java/com/hybridforge/training/datamodel/Person.java
package com.hybridforge.training.datamodel;

import java.util.Arrays;
import java.util.List;

/**
 * Models a person and their age
 *
 * Created by jbenito on 14-12-30.
 */
public class Person {
    private String mFirstName;
    private String mLastName;
    private int mAge;

    static List<Person> mTestData1;
    static List<Person> mTestData2;
    static Boolean mTag;

    static {
        Person[] sampleData = new Person[] { Person.newInstance("Joe", "Brown", 25),
                Person.newInstance("Walter", "White", 27),
                Person.newInstance("Jesse", "McCartney", 25),
                Person.newInstance("George", "Bush", 12),
                Person.newInstance("Barack", "Obama", 29)};
        mTestData1 = Arrays.asList(sampleData);

        sampleData = new Person[] { Person.newInstance("Daffy", "Duck", 25),
                Person.newInstance("Mickey", "Mouse", 27),
                Person.newInstance("Roger", "Rabbit", 25),
                Person.newInstance("Bugs", "Bunny", 12),
                Person.newInstance("Speedy", "Gonzales", 29)};
        mTestData2 = Arrays.asList(sampleData);

        mTag = false;
    }

    /**
     * Default constructor
     * @param firstName Person's first name
     * @param lastName Person's last name
     * @param age Person's age
     */
    public Person(String firstName, String lastName, int age) {
        mFirstName = new String(firstName);
        mLastName = new String(lastName);
        mAge = age;
    }

    /**
     * Convenience method for creating a new instance of Person
     * @param firstName Person's first name
     * @param lastName Person's last name
     * @param age Person's age
     * @return a new instance of Person with the specified input attributes
     */
    public static Person newInstance(String firstName, String lastName, int age) {
        return new Person(firstName, lastName, age);
    }

    /**
     * Convenience method for creating a
     * @return
     */
    public static List<Person> getTestData() {
        if (mTag) {
            return mTestData1;
        }
        return mTestData2;
    }

    public static void toggleTestData() {
        mTag = !mTag;
    }

    /**
     * Returns the concatenation of the first name and last name
     * @return Person's full name
     */
    public String getFullName() {
        return String.format("%s %s", mFirstName, mLastName);
    }

    /**
     * Returns the age in a displayable format
     * @return e.g. "1 year old", "5 years old", "-"
     */
    public String getAgeString() {
        if (mAge > 1) {
            return String.format("%d years old", mAge);
        }
        if (mAge == 1) {
            return "1 year old";
        }
        return "-";
    }

    @Override
    public int hashCode() {
        return getFullName().hashCode() + getAgeString().hashCode();
    }
}

File: iOS/Solution/Exercise3_part2/Training/Model/Person.m
//
//  Person.m
//  Training
//
//  Created by JD Benito on 2015-01-05.
//  Copyright (c) 2015 Hybrid Forge. All rights reserved.
//

#import "Person.h"

@interface Person()
- (id)initWithFirstName:(NSString *)firstName_ lastName:(NSString *)lastName_ age:(NSInteger)age_;
@end

@implementation Person

+ (NSArray *)testData1 {
    return @[ [Person personWithFirstName:@"Joe" withLastName:@"Brown" andAge:25],
              [Person personWithFirstName:@"Walter" withLastName:@"White" andAge:27],
              [Person personWithFirstName:@"Paul" withLastName:@"McCartney" andAge:25],
              [Person personWithFirstName:@"George" withLastName:@"Bush" andAge:12],
              [Person personWithFirstName:@"Barack" withLastName:@"Obama" andAge:29] ];
}

+ (NSArray *)testData2 {
    return @[ [Person personWithFirstName:@"Daffy" withLastName:@"Duck" andAge:25],
              [Person personWithFirstName:@"Mickey" withLastName:@"Mouse" andAge:27],
              [Person personWithFirstName:@"Roger" withLastName:@"Rabbit" andAge:25],
              [Person personWithFirstName:@"Bugs" withLastName:@"Bunny" andAge:12],
              [Person personWithFirstName:@"Speedy" withLastName:@"Gonzales" andAge:29] ];
}

+ (Person *)personWithFirstName:(NSString *)firstName_ withLastName:(NSString *)lastName_ andAge:(NSInteger)age_ {
    return [[Person alloc] initWithFirstName:firstName_ lastName:lastName_ age:age_];
}

- (id)initWithFirstName:(NSString *)firstName_ lastName:(NSString *)lastName_ age:(NSInteger)age_ {
    self = [super init];
    if (self != nil) {
        _firstName = [firstName_ copy];
        _lastName = [lastName_ copy];
        _age = age_;
    }
    return self;
}

- (NSString *)fullName {
    return [NSString stringWithFormat:@"%@ %@", _firstName, _lastName];
}

- (NSString *)displayAge {
    if (_age > 1) {
        return [NSString stringWithFormat:@"%ld years old", (long)_age];
    }
    if (_age == 1) {
        return @"1 year old";
    }
    return @"-";
}

@end
